python/paddle/fluid/dygraph/dygraph_to_static/list_transformer.py

```diff
@@ -0,0 +1,169 @@
+
+    def _need_to_array_write_node(self, node):
+        if isinstance(node, gast.Expr):
+            if isinstance(node.value, gast.Call):
+                if self._is_list_append_tensor(node.value):
+                    return True
+
+        return False
+
+    def replace_list_with_tensor_array(self, node):
+        for child_node in gast.walk(node):
+            if isinstance(child_node, gast.Assign):
+                if self._need_to_create_tensor_array(child_node):
+                    child_node.value = self._create_tensor_array()
+
+    def _is_list_append_tensor(self, node):
+        """
+        a.append(b): a is list, b is Tensor
+        """
+        assert isinstance(node, gast.Call)
+        # 1. is .append
+        if not isinstance(node.func, gast.Attribute):
+            return False
+        if node.func.attr != 'append':
+            return False
+
+        # 2. is list
+        value_name = astor.to_source(gast.gast_to_ast(node.func.value))
+        if value_name.strip() not in self.list_name_to_updated:
+            return False
+
+        # 3. append Variable
+        # Only one argument is supported in Python list.append()
+        if len(node.args) != 1:
+            return False
+
+        arg = node.args[0]
+        if isinstance(arg, gast.Name):
+            # TODO: `arg.id` may be not in scope_var_type_dict if `arg.id` is the arg of decorated function
+            # Need a better way to confirm whether `arg.id` is a Tensor.
+            try:
+                var_type_set = self.scope_var_type_dict[arg.id]
+            except KeyError:
+                return False
+
+            if NodeVarType.NUMPY_NDARRAY in var_type_set:
+                return False
+            if NodeVarType.TENSOR not in var_type_set and NodeVarType.PADDLE_RETURN_TYPES not in var_type_set:
+                return False
+        # else:
+        # Todo: Consider that `arg` may be a gast.Call about Paddle Api.
+        # eg: list_a.append(fluid.layers.reshape(x))
+        # return True
+        self.list_name_to_updated[value_name.strip()] = True
+        return True
+
+    def _need_to_create_tensor_array(self, node):
+        assert isinstance(node, gast.Assign)
+        target_node = node.targets[0]
+        try:
+            target_id = target_node.id
+        except AttributeError:
+            return False
+        if self.list_name_to_updated.get(target_id):
+            return True
+        return False
+
+    def _create_tensor_array(self):
+        # Although `dtype='float32'`, other types such as `int32` can also be supported
+        func_code = "fluid.layers.create_array(dtype='float32')"
+        func_node = gast.parse(func_code).body[0].value
+        return func_node
+
+    def _to_array_write_node(self, node):
+        assert isinstance(node, gast.Call)
+        array = astor.to_source(gast.gast_to_ast(node.func.value))
+        x = astor.to_source(gast.gast_to_ast(node.args[0]))
```

Comment:
I add a common function `ast_to_source_code` in `utils.py` in PR https://github.com/PaddlePaddle/Paddle/pull/22967. Codes here can be replaced by `ast_to_source_code` in the future to make it clearly.

Reply:
Great. Code will be changed to `ast_to_source_code`